BUG: group with multiple named results

guenteru opened the pull request:
- [x] closes #19029 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This bugfix gets rid of duplicated names that can be the result of groupby operations (#19029).
I opted to implement one of the ideas proposed by [toobaz](https://github.com/pandas-dev/pandas/issues/19029#issuecomment-354684340): duplicated names get suffixed by their corresponding position, i.e. ['name','name'] gets transformed into ['name0', 'name1'].

* A few testcases have been added.
* One particular testcase had to be changed ([test_crosstab_dup_index_names](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/reshape/test_pivot.py))
  - This is because with the new bugfix crosstab does not yield a ValueError anymore.

guenteru committed:
add fix for bug 19029

As of version 0.23.0 MultiIndex throws an exception in case it contains
duplicated level names. This can happen as a result of various groupby
operations (21075). This commit changes the behavior of groupby slightly: In
case there are duplicated names contained in the index these names get suffixed by there
corresonding position (i.e. [name,name] => [name0,name1])

guenteru committed:
update old testcase to satisfy new behavior

guenteru committed:
add additional groupby testcases (19029)

pep8speaks commented:
Hello @guenteru! Thanks for submitting the PR.

 - In the file [`pandas/core/groupby/groupby.py`](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py), following are the PEP8 issues :

> [Line 2303:21](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py#L2303): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 2304:62](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py#L2304): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 2309:42](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py#L2309): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 2310:33](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py#L2310): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 2310:38](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/core/groupby/groupby.py#L2310): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

 - In the file [`pandas/tests/groupby/test_categorical.py`](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_categorical.py), following are the PEP8 issues :

> [Line 562:11](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_categorical.py#L562): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 567:19](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_categorical.py#L567): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 568:1](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_categorical.py#L568): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 569:1](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_categorical.py#L569): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

 - In the file [`pandas/tests/groupby/test_groupby.py`](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py), following are the PEP8 issues :

> [Line 1680:80](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1680): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1681:80](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1681): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1693:19](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1693): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 1695:34](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1695): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1700:80](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1700): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 1709:25](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1709): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1713:19](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1713): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 1715:34](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1715): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1716:1](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/groupby/test_groupby.py#L1716): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

 - In the file [`pandas/tests/reshape/test_pivot.py`](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py), following are the PEP8 issues :

> [Line 1712:12](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1712): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 1712:18](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1712): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1712:32](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1712): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1716:23](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1716): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to False should be 'if cond is False:' or 'if not cond:'
> [Line 1718:32](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1718): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1719:32](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1719): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1720:55](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1720): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1720:80](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1720): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 1721:37](https://github.com/pandas-dev/pandas/blob/32e44c34f73a5224a624740a051d61b969a274a2/pandas/tests/reshape/test_pivot.py#L1721): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

guenteru committed:
resolve flake8 conflicts

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -2298,7 +2298,18 @@ def levels(self):
+            return orig_names
+        # in case duplicates are contained rename all of them
+        if len(set(orig_names)) < len(orig_names):
+            orig_names = [''.join([str(x), str(i)])
```
Use `.format` instead of `.join`. I'd also suggest adding an underscore like we do in merges

WillAyd commented on pandas/core/groupby/groupby.py:
```diff
@@ -2298,7 +2298,18 @@ def levels(self):
     def names(self):
-        return [ping.name for ping in self.groupings]
+        # GH 19029
+        # add suffix to level name in case they contain duplicates (GH 19029):
```
Can remove ref to GH here, since it's on the line above

WillAyd commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -558,9 +558,13 @@ def test_as_index():
-    # GH18872: conflicting names in desired index
-    with pytest.raises(ValueError):
+    # GH 19029: conflicitng names should not raise a value error anymore
+    raised = False
```
Rather than doing this you should say what the `expected` object is and compare to that the result of the groupby

WillAyd commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1674,3 +1674,42 @@ def test_tuple_correct_keyerror():
+
+def test_dup_index_names():
+    # dup. index names in groupby operations should be renamed (GH 19029):
+    df = pd.DataFrame({'date': list(pd.date_range('5.1.2018', '5.3.2018')),
```
Don't think you need to wrap the date_range in a list (?)

WillAyd commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1674,3 +1674,42 @@ def test_tuple_correct_keyerror():
+
+    failed = False
+    try:
+        result = df.groupby([df.date.dt.month, df.date.dt.day])['vals'].sum()
```
Same comment as before - don't need this try...catch block. For what it's worth though let's not duplicate tests, so either have this check here in its own method and remove from the other test, or just update the other test and leave this. See which one makes more sense

WillAyd commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1674,3 +1674,42 @@ def test_tuple_correct_keyerror():
+    tm.assert_series_equal(result, expected)
+
+
+def test_empty_index_names():
```
What happens if say 2 out of 3 levels use `None` as a name - would these tests still hold?

guenteru committed:
change groupby-behaviour (duplicates) & tests

Only duplicates get suffixed by their corresponding enumeration value:
['name', None, 'name'] gets transformed into ['name_0', None, 'name_1']

Superfluous test cases have been deleted and some additonal test
statements have been added.

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -2298,7 +2298,28 @@ def levels(self):
     @property
     def names(self):
-        return [ping.name for ping in self.groupings]
+        # add suffix to level name in case they contain duplicates (GH 19029):
```
this is super complicated, what exactly are you trying to do here?

guenteru replied:
Hi. The goal here is to add a suffix to duplicate entries:
```python
# takes care of multiplicity:
['x', 'x', 'y', 'y']  => ['x_0', 'x_1', 'y_0', 'y_1']
```
Before the current version the code just enumerated all the entries regardless of their multiplicity:
```python
['x', 'x', 'y', 'y']  => ['x_0', 'x_1', 'y_2', 'y_3']
```
For some reason I thought that the current version would be better suited.
I can switch back to the old (and probably less confusing) version if you want.

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -1674,3 +1674,44 @@ def test_tuple_correct_keyerror():
+
+
+def test_dup_index_names():
+    # dup. index names in groupby operations should be renamed (GH 19029):
```
can you parameterize this

jreback commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -1705,9 +1705,21 @@ def test_crosstab_with_numpy_size(self):
         s = pd.Series(range(3), name='foo')
-        pytest.raises(ValueError, pd.crosstab, s, s)
+        failed = False
+        try:
```
if you are asserting that this works, simply collect the result and compare vs expected

jreback commented:
closing as stale. if you'd like to continue pls ping.
